bukuserver/api.py

```diff
@@ -158,45 +185,52 @@ def delete(self, rec_id: T.Union[int, None]):
         else:
             bukudb = getattr(flask.g, 'bukudb', get_bukudb())
             result_flag = bukudb.delete_rec(rec_id)
-        return to_response(result_flag)
+        return Response.from_flag(result_flag)
 
 
 class ApiBookmarkRangeView(MethodView):
 
     def get(self, starting_id: int, ending_id: int):
         bukudb = getattr(flask.g, 'bukudb', get_bukudb())
         max_id = bukudb.get_max_id() or 0
-        if starting_id > max_id or ending_id > max_id:
-            return response_bad()
+        if starting_id > ending_id or ending_id > max_id:
+            return Response.FAILURE()
         result = {'bookmarks': {i: entity(bukudb.get_rec_by_id(i))
                                 for i in range(starting_id, ending_id + 1)}}
-        return jsonify(result)
+        return Response.SUCCESS(data=result)
 
     def put(self, starting_id: int, ending_id: int):
         bukudb = getattr(flask.g, 'bukudb', get_bukudb())
         max_id = bukudb.get_max_id() or 0
-        if starting_id > max_id or ending_id > max_id:
-            return response_bad()
-        for i in range(starting_id, ending_id + 1, 1):
-            updated_bookmark = request.data.get(str(i))  # type: ignore
-            result_flag = bukudb.update_rec(
-                i,
-                updated_bookmark.get('url'),
-                updated_bookmark.get('title'),
-                updated_bookmark.get('tags'),
-                updated_bookmark.get('description'))
-            if result_flag is False:
-                return response_bad()
-        return response_ok()
+        if starting_id > ending_id or ending_id > max_id:
+            return Response.FAILURE()
+        for rec_id in range(starting_id, ending_id + 1):
+            data = request.get_json().get(str(rec_id))
+            form = ApiBookmarkEditForm({})
+            form.process(data=data)
+            if not form.validate():
+                return Response.INPUT_NOT_VALID()
+            if data.get('tags') is None:
+                tags_str = None
+            else:
+                tags_str = buku.parse_tags([buku.DELIM.join(form.tags.data)])
```

Comment:
…I'd say turning a fairly simple expression into a 4-line block is a bit of an overkill :sweat_smile:

Reply:
…Say. Does `form.tags.data` contain a value different from `data.get('tags')` when the latter is neither a string nor `None`? Because if not, it'd make sense to use the `tags` binding you added above in place of `form.tags.data` here.

Reply:
Come to think of it, if `data` is `None`, `data.get('tags')` would cause an `AttributeError` (and in the current version this call happens before form validation, so I'm guessing this would result in HTTP 500?).

Reply:
yes if `form.validate()` returns True (i.e. there is no comma in strings), then `form.tags.data` should be the same as tags. However I pass form data for all other fields so i think it'd make sense to be consistent.

I suggest creating a [custom FieldList](https://wtforms.readthedocs.io/en/3.0.x/fields/#custom-fields) and let the form handle the processing from list to string

Reply:
> Come to think of it, if `data` is `None`, `data.get('tags')` would cause an `AttributeError` (and in the current version this call happens before form validation, so I'm guessing this would result in HTTP 500?).

`data.get('tags') is short for data.get('tags', None) so it will return None if 'tags' field does not exist, not AttributeError

Reply:
Sorry i misread you, you meant `data=None`. I think if the request if properly formed (i.e. json is passed) `request.get_json()` should not return None

Reply:
…And maybe the `TAGS_NOT_LIST` check should be actually checking for the `tags` value being a _list_ (as opposed to _string_)?

I also suspect that this is checked by the form validation.

…Speaking of which: shouldn't the request data be validated _before_ we start applying DB changes? As of right now, it seems that if the request is malformed, we end up applying some part of it before reporting a failure.

Reply:
Yes the form is responsible for checking that the value associated with tags is a list. The reason why I first check that tags is not a string is because the form will still pass validation and treat the string value as a list of characters... For example when passing {`tags`: `not a list`} I get `tags_str = ,a,i,l,n,o,s,t`

True, I could validate all forms before proceeding.

Reply:
> `request.get_json()` should not return `None`

Technically `null` is a valid JSON that's equivalent to `None`. But the case I'm talking about is actually `get_json().get(…)` returning `None`.

Reply:
> I suggest creating a [custom FieldList](https://wtforms.readthedocs.io/en/3.0.x/fields/#custom-fields) and let the form handle the processing from list to string

That _is_ an option, I suppose… So long as it doesn't outright break on invalid values nor treat `None` as `[]` where it shouldn't.

Reply:
So it turns out that `FieldList` was not actually checking that the value passed to `tags` was actually a list. What it does is it [loops through the data being passed](https://github.com/wtforms/wtforms/blob/master/src/wtforms/fields/list.py#L92) and fills its subfields (hence if data is a string it will loop through characters). Subfields are validated first, then FieldList validators are run (in this case there aren't any).  Adding a custom validator to `FieldList` did not help because the data was already processed when it reaches the validator (i.e. `None` resolves to `[]` and strings already split into a list of chars).
Instead I decided to create a custom `process_data` method in the form that processes the input, validates the form and parse the tags string for later use.

Reply:
…Make sure to double-check the API manually if you haven't yet (just in case).

Meanwhile, I'll be trying to figure out how to fix broken tests :sweat_smile: